src/Tools/AI Test Toolkit/src/Logs/AITRunHistory.Page.al

```diff
@@ -0,0 +1,222 @@
+                    {
+                        Visible = ViewBy = ViewBy::Tag;
+                        Caption = 'No. of Tests Passed';
+                        ToolTip = 'Specifies the number of tests passed for the tag.';
+                        Style = Favorable;
+                    }
+                    field("No. of Tests Failed - By Tag"; Rec."No. of Tests Executed - By Tag" - Rec."No. of Tests Passed - By Tag")
+                    {
+                        Visible = ViewBy = ViewBy::Tag;
+                        Editable = false;
+                        Caption = 'No. of Tests Failed';
+                        ToolTip = 'Specifies the number of tests that failed for the tag.';
+                        Style = Unfavorable;
+
+                        trigger OnDrillDown()
+                        var
+                            AITLogEntryCU: Codeunit "AIT Log Entry";
+                        begin
+                            AITLogEntryCU.DrillDownFailedAITLogEntries(Rec."Test Suite Code", Rec."Line No. Filter", Rec.Tag);
+                        end;
+                    }
+                    field("Duration - By Tag"; Rec."Total Duration (ms) - By Tag")
+                    {
+                        Visible = ViewBy = ViewBy::Tag;
+                        Caption = 'Total Duration (ms)';
+                        ToolTip = 'Specifies Total Duration of the tests for the base version.';
+                    }
+                    field("Tokens - By Tag"; Rec."Tokens Consumed - By Tag")
+                    {
+                        Visible = ViewBy = ViewBy::Tag;
+                        Caption = 'Total Tokens Consumed';
+                        ToolTip = 'Specifies the aggregated number of tokens consumed by the test in the current version. This is applicable only when using Microsoft AI Module.';
+                    }
+                }
+            }
+        }
+    }
+
+    var
+        TestSuiteCode: Code[100];
+        ViewBy: Enum "AIT Run History - View By";
+        LineNo: Integer;
+        ApplyLineFilter: Boolean;
+        LineNoFilter: Text;
+
+    trigger OnOpenPage()
+    begin
+        UpdateRunHistory();
+    end;
+
+    internal procedure SetTestSuite(Code: Code[100])
+    begin
+        TestSuiteCode := Code;
+    end;
+
+    internal procedure FilterToLine(Line: Integer)
+    begin
+        ApplyLineFilter := true;
+        LineNo := Line;
+    end;
+
+    local procedure LookupTestMethodLine()
+    var
+        AITLogEntry: Codeunit "AIT Log Entry";
+    begin
+        AITLogEntry.LookupTestMethodLine(TestSuiteCode, LineNoFilter, LineNo);
+        UpdateRunHistory();
+    end;
+
+    local procedure UpdateRunHistory()
+    var
+        AITLogEntry: Codeunit "AIT Log Entry";
+    begin
+        if not ApplyLineFilter then
+            LineNo := 0;
+
+        AITLogEntry.UpdateRunHistory(TestSuiteCode, LineNo, ViewBy, Rec);
```

Comment:
Probably just do filter here after Get or before, what with the LineNo. Then remove it as param.

Reply:
Again, I think we should centralize code in codeunits when possible.